index.rst

```diff
@@ -1,51 +1,1667 @@
+        clsInterpolate.def("interpolate",
+                           (ndarray::Array<double, 1> (Interpolate::*) (ndarray::Array<double const, 1> const&)
+                               const) &Interpolate::interpolate);
+
+        return mod.ptr();
+    }
+
+Other Useful Tips
+=================
+
+Operators
+---------
+
+You may find it necessary to wrap operators.
+While pybind11 contains a useful syntax to easily wrap operators,
+we have found that it doesn't work as often as one would like.
+Instead, we wrap an operator with a lambda function,
+for example to overload the multiplication operator for a class A we use
+
+.. code-block:: c++
+
+    cls.def("__mul__", [](A const & self, A const & other) {
+        return self * other;
+    }, py::is_operator());
 
 .. note::
 
-   **This technote is not yet published.**
+    The ``py::is_operator()`` informs pybind11 that the wrapped function is an operator which should
+    trigger a ``NotImplementedError`` instead of a ``TypeError`` when called with the wrong type.
+
+.. _python-code:
+
+Python Code
+-----------
+
+In some cases C++ classes are extended to include methods specific to the python API,
+or to make C++ objects and methods more pythonic.
+Unlike SWIG, which has a specific ``extend`` method,
+monkey-patching like this is frowned upon in python and no formal method exists to extend a C++ class.
+The following example provides the recommended method for extending C++ classes in our stack.
+
+``afw::table`` contains an ``Arrays.h`` header file that defines the
+``ArrayFKey`` and ``ArrayIKey`` objects.
+The relevant pybind11 wrapper code ``arrays.cc`` is shown below:
+
+.. code-block:: c++
+
+    template <typename T>
+    void declareArrayKey(py::module & mod, std::string const & suffix) {
+        py::class_<ArrayKey<T>,
+                   std::shared_ptr<ArrayKey<T>>,
+                   FunctorKey<ndarray::Array<T const, 1, 1>>> clsArrayKey(mod, ("Array"+suffix+"Key").c_str());
+    
+        clsArrayKey.def(py::init<>());
+        clsArrayKey.def("_get_", [](ArrayKey<T> & self, int i) {
+            return self[i];
+        });
+        clsArrayKey.def("getSize", &ArrayKey<T>::getSize);
+        clsArrayKey.def("slice", &ArrayKey<T>::slice);
+    };
+
+    PYBIND11_PLUGIN(_arrays) {
+        py::module mod("_arrays", "Python wrapper for afw _arrays library");
+    
+        if (_import_array() < 0) {
+                PyErr_SetString(PyExc_ImportError, "numpy.core.multiarray failed to import");
+                return nullptr;
+        };
+
+        /* Module level */
+        declareArrayKey<float>(mod, "F");
+        declareArrayKey<double>(mod, "D");
+
+        return mod.ptr();
+    }
+
+In this case it is useful to make the ``get`` method in
+``ArrayFKey`` and ``ArrayDKey`` more pythonic by allowing them to
+accept slices as well as indices, so we create a new file ``arrays.py``
+(notice the difference between this and ``_arrays.py``, which is created by pybind11)
```

Comment:
There is no `_arrays.py`. There's an `_arrays.so`, but maybe it would be more useful to talk about the `_arrays` module.